README
yes, I am insane

Tired of waiting many miliseconds for linux_logo to run?
Tired of wasting 35k of disk space?
Upset that to run linux_logo you need huge GLIBC?

Well your worries are over!

With "ll"  [even the name was shortened to save space!] you get all of the
benefits of linux_logo in a smaller, faster package!

"ll" is written entirely in native Linux assembly language!


Some Statistics
---------------

     *NOTE*  not all architectures implement the same feature-set
       (IE, not all have MHz in /proc/cpuinfo) so this is only
       a rough comparison.


     Processor        lzss executable
     ---------        ---------------
     ia64		2826 bytes   (as 2.18.0.20080103)
     alpha		1821 bytes   (as 2.19.51.20090723)
     RiSC		1418 bytes   (RiSC-as 0.0.2)
     parisc		1400 bytes   (as 2.17.50)
     SPARC		1397 bytes   (as 2.19.51.20090714)
     microblaze		1298 bytes   (as 2.16)
     mips               1276 bytes   (as 2.19.51.20090714)
     m88k		1240 bytes   (as 1.92.3)
     arm.eabi		1186 bytes   (as 2.22)
     arm.oabi		1186 bytes   (as 2.18.0.20080103)
     PPC		1165 bytes   (as 2.19.51.20090805)
     6502		1130*bytes   (ca64 2.12.0)
     arm64		1094 bytes   (as 2.23.1)
     s390	        1064 bytes   (as 2.19.51.20090805)
     x86_64		1030 bytes   (as 2.23.52.20130727)
     x86_x32            1006 bytes   (as 2.24)
     sh3		 994 bytes   (as 2.17.50.0.5)
     m68k		 982 bytes   (as 2.18.0.20080103)
     x86                 969 bytes   (as 2.19.51.20090723)
     vax		 950 bytes   (as 2.16.1)
     arm_thumb		 937 bytes   (as 2.22)
     arm_thumb2		 925 bytes   (as 2.22)
     avr32		 914 bytes   (as 2.16.1)
     crisv32		 905 bytes   (as 2.12.1)
     z80		 891 bytes   (as 2.20.1.20100303)
     pdp-11		 890 bytes   (as 2.19)
     8086		 780 bytes   (as 2.20.1-system.20100303)
     
     * the 6502 results were adjusted to match the code present in other
       architectures (i.e., not counting the graphical routines)
     
     The various implementations have varying functunality and often
     use different methods to get system info.  Still, some gross
     comparisons between the architectures can be made.


Individual architectural comments, in descending order of executable size:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ia64:   ia64 is VLIW.
     
	no divide at all.
	you use the fp unit to do int multiply
	no unaligned words.
	on the plus side, basically have infnite (well, 128) registers

	does have an auto-incrementing load/store.
	
	The actual ia64 architecture is too bizzarre for words.
	It probably doesn't make any sense at all unless you are a 
	   computer architect.
	The instructions come in groups of 3 (40-bits each, with
	   a total instruction size of 128bits).  These run in
	   parallel.  So if your instructions don't parallelize,
	   you end up running lots of nops.
	   
	In any case it's no suprise the executable ends up being
	   so huge.  It very probably can be optimized a lot from here.
	If I really cared I'd turn off the automatic bundling
	   and set all of the instruction bundles by hand.

	   I found a bug in the assembler where it puts two
	   call instructions in the same bundle (which can't work).
	   I wonder if thise means I am one of the few people who
	   writes programs entirely in assembly for ia64....

alpha:  
	Alpha is hurt because it implements some "optional" features,
	such as MHz (which added a lot of code) and also counting
	num of cpus, with proper pluralization of Processor.
	
	The *big* hurt though is lack of byte-manipulating instructions.

	The original alpha architecture did not support operations on bytes,
	you have to do a lot of shifting and masking.
	
	Unfortunately ll uses a lot of byte-sized memory operations.
     
        On the x86 the instruction "lodsb" is 1 byte in size.
        On the PPC the equivelant is "lbzu" which is 4 bytes in size.
        On the alpha, the instruction is "ldb" which expands to
           a "lda","ldq_u","extbl","sll","sra" sequence, plus
	   an add instruction that x86 and ppc do automatically.
	   Thus, taking 24 bytes.
	It's a bit better if you do an unsigned load, which is only 16
	   bytes, but still.

	A store byte actually does a 32-bit load, masks in the byte by
	  hand, and then does the actual 32-bit store :(

	The "jump if bit one set" type instructions help the lzss.

	The immediate field for ALU ops of only 8 bits really hurts
	
	There is no native integer division routine on alpha.
	
	The GOT section is a huge space-sink.  It has 64-bit constants,
	  which most of the time you shouldn't really need.

RiSC:
	This is a 16-bit architecture used in my undergrad
	computer courses I took at University of Maryland.
		 http://www.ece.umd.edu/~blj/RiSC/
		 
	No logic instructions besides nand.  This means
	masking with an "and" takes at least 2 instructions 
	(because there is no immediate form).
	
	16-bit memory acesses only.  It is quite complicated to load a byte.
	
	Only 7 registers, one of which is always 0, and one of which
	is the stack pointer.
	
	No shift instructions.  Left shift can be done with add, but
	right shift takes a fairly complicated routine.

        Only able to branch +/- 64 instructions, which can be a limit.
	
	Only a 6-bit immediate, though the lui instruction helps.



parisc:	+ really hurt by its short immediate field.  Most addresses
	  require 2 instructions to load, even if try to use relative-data add.
	+ no integer divide, have to code it up... not so bad in loop form
        + delay slot that can be nulled out
	+ some ALU instructions can also null out following instructions 
          conditionally
	+ compare immediate instruction only can handle 5-bit immediate	
        + loads/stores must be aligned
	+ no AND immediate instruction

sparc:  
	Condition codes make for tighter code.
	register/register load address calcs also help.
        13-bit immdediate hurts a bit.
        Crazy register windows a bit hard to understand

	SPARC is unfair a bit, because my test machine is a 24-proc
        niagara, so it has extra code to handle that many chips properly.

65c816:
        Another non-Linux addition.
	
	The method of switching the A and X, Y registers into and out
	of 16 bit mode is a PAIN.  A non-intuitive opcode, and if you
	aren't careful your routines break if you're in an unexpected mode.
	Would have been much better if somehow there were separate opcodes
	for 8 vs 16-bit instructions.

6502:
        So obviously this isn't running on Linux.  I was curious
	how an 8-bit processor would compare.
	
	The big problem is that the LZSS algorithm and the ll data
	set are very much 16-bits in size.  So there is a lot of
	overhead having to increment 16-bit values on an 8-bit processor.
	
	Only having 3 registers is a handicap, but the zero page (the
	first 256 bytes of memory which can be accesses in one less byte
	and in fewer cycles) act as almost virtual registers.
	
	Some potential useful instructions that would have helped (and
	that are actually implemented in the later 65C02 version of the
	chip):
	   phx/plx (push/pull X directly)
	   ina/dea (increment/decrement A... otherwise need 2 instructions)
	   bra (branch always, relative jump.  otherwise need 3-byte 16-bit)
	   stz (store zero)

        The 6502 code uses high-res graphics to approximate the color
	ascii art.  When calculating total image size, we don't count
	the overly-large graphics code and instead only count the code
	that would be used to print the text, say, out over a serial port
	to an ANSI capable term client.
	
microblaze:
        Similar to MIPS.
	
	Branch delay slot is optional, which helps.
	
	Only register+register and register+immediate address modes.
	
	The branch+link instruction returns to the same instruction
	  that left from, so for the return you have to specify
	  to add 4 or 8?  Very weird.
	  
	The signed 16-bit immediate makes it impossible (I think) to load
	  0xff00 into a register w/o two instructions.

        Can't get the assembler to do pointer math into an immediate.

mips:   Recent binutils has made mips come in line with the
        other architectures.
	
	It is the most RISC of the RISC architectures.  Thus it ends
	up having a very non-dense instruction set.
	
	On the plus side, it has hardware support for unaligned loads,
	plus hardware integer divide, which help a lot.

m88k:	This chip came after m68k and was motorola's first RISC chip.
	Didn't go well, they moved on to PPC.
	
	It's similar to a cross between MIPS and PPC.
	
	OpenBSD because there is no Linux port to m88k.
	Using gxemul.
	
	OpenBSD forces syscall error handling by skipping a 
	jump-to-error-handler instruction on return.  This
	makes each syscall in effect take 8 bytes.
	
	Branch delay slot is optional (for performance only?)
	
	Starts at low address space.  This is a huge win as it
	makes each pointer load 4-bytes instead of 8 assuming
	we stay smaller than 64k.

	bcnd instruction saves a cmp usually
	
	cmp quite elaborate, setting 16-bits of comparison info in
	a reg, not in flags.  a more powerful version of the equivelant
	Alpha instruction.

arm64:
	New ISA different from arm32/thumb/thumb2
	Fixed-width 32-bit instructions

	Wider constants help *a lot*
	Unaligned loads help too
	New instructions, such as tbz also help
	The conditional instructions like csel might, but at least in the
		lzss code doesn't make a difference.	
	The compare/branch/zero instructions also help.

	Lack of conditional execution hurts

	What we really need is an "increment multiple" instruction.
		inc {r1,r2,r4}
	And also an equivelent of x86 loop	
	
	No open syscall, have to waste an instruction setting
	AT_FDCWD and use openat() instead.

arm:	no integer division routine
	Really painful to load constants > 8 bits that aren't powers of two or

CHANGES
?? 2001
+ Got interested in linux-assembly when working at Frontpath.
+ First x86 version written in this time frame.
+ The assembly code was distributed as part of the linux_logo
  code at first.

5 March 2002
+ Start work on the ppc version

8 March 2002
+ RLE (run-length-encode) compressed the logo.  Cuts 1.3k off the
  size of the intel binary

11 March 2002
+ Convert from nasm to 'gas' as the assembler.  Yes, I know, ick, ick, ick,
  but it wasn't as bad as I thought it'd be.
+ added "configure.c" which will detect architecture and compile correct
  assembly file

13 March 2002
+ Finally have ppc version all working

6 May 2002
+ Add ansi_rle.c program which will convert any ansi into a form
  suitable to use with ll
+ Re-wrote the code a bit to handle generic ANSI sequences, not just
  color-changes anymore

7 May 2002
+ Make 'll' return to normal plain console color after running

RELEASE 0.5

5 June 2002
+ Add Alpha version.  Thanks to the compaq test drive people

RELEASE 0.6

8 June 2002
+ Add ia64 version.  Thanks again to compaq test drive.

RELEASE 0.7

2 July 2002
+ Add SPARC version

RELEASE 0.8

26 July 2002
+ Import "configure" from linux_logo-4.03
+ Optimize 208bytes off of ix86 code, thanks to suggestions from
  Stephan Walter <stephan.walter _at_ gmx.ch>
+ Add the RiSC port to the 16bit directory  

RELEASE 0.9

2 August 2002
+ added s390 support

RELEASE 0.10

8 August 2002
+ Add "lzss" logo support and convert ix86 over to new code
  (with much help from Stephan Walter).

RELEASE 0.11

22 August 2002
+ Have ix86 code down to 1476 bytes!
+ Have partially working PPC version of lzss code

28 August 2002
+ Update s390 to lzss before my testdrive account expires
+ Update ix86 comments a bit

10 September 2002
+ re-written PPC code is 1796 bytes!
+ release 0.12

RELEASE 0.12

27 September 2002
+ added "sstrip" utility by Brian Raiter which really reduces the size
+ got the ix86 code to 1021 bytes!!! thanks to Stephan Walter again,
  but also I re-wrote the lzss core and did some tricks myself

RELEASE 0.13

16 December 2002
+ Thanks to Stephan Walter again for getting the ix86 code down to
  993 bytes.
* Worked on the code myself and got it down to 977 by using edx as a
  function pointer and using 2-byte call *%edx.  The ix86 code
  is becoming so optimized as to be unreadable...
  
RELEASE 0.14

13 January 2006
+ Mips version working.  Should optimize for size a bit. 
  It is based off of the x86 code so uses the more optimized
  LZSS logo.
+ Work on the PPC version which stopped working, apparently
  because the "ctr" register is not saved across a syscall with
  kernel 2.6.  Interesting.
  
19 May 2006
+ Start work on the HP PA-RISC version
+ Fix up the infrastructure to special-case the weird .EQU
  directive order the hppa arch needs.
  
13 January 2007
+ Work on PA-RISC again after long delay.  Have it working, but not optimized
  
RELEASE 0.15
RELEASE 0.16

29 January 2007
+ Optimized PA-RISC code.  sstrip'd version before was 2096 bytes, after is
  1432 bytes.

RELEASE 0.17

2 February 2007
+ Change the way the Makefile detects architecture
+ x86 ll was crashing on newer machines with really stupid cpuinfo cpu types
  (core2 and newer xeons, I'm looking at you).  The problem was it makes
  the middle string >80, which the check for was broken.  Fix, but it
  adds an extra 3 bytes to program length.

3 February 2007
+ x86_64 version.  Hoped I could make it smaller, but for various reasons
  it ends up being a tens of bytes bigger than x86 version.  Try 
  various optimizations to no avail.

RELEASE 0.18

6 February 2007
+ Re-optimize the mips version.  It is now 160 bytes smaller.

RELEASE 0.19

9 February 2007
+ Fix branch-delay slot optimization in strcat in MIPS.
  Binary now 16 bytes smaller!
+ Re-optimize SPARC version.
+ Added a FAQ.

RELEASE 0.20

16 February 2007
+ Start optimizing Alpha code

17 February 2007
+ Finish optimizing Alpha code.  It could use another pass though.

RELEASE 0.21

27 February 2007
+ Have ARM completely working.
+ Started on arm thumb version, but on my development board the kernel isn't
  compiled with thumb support, and I'm a bit wary of flashing the boot flash.
  I'll try to get a kernel booted off a CF card at some point.

RELEASE 0.22

11 March 2007
+ Thumb support working.  Wasted a day chasing down an off-by-one
  loop thinko with strcat.
+ Optimized thumb support.
  
RELEASE 0.23
  
22 March 2007
+ Update itanium code to use LZSS logo.  It's the last of the RLE
  executables.

RELEASE 0.24

24 April 2007
+ Get aranym going for m68k emulation.

26 April 2007
+ Initial m68k work

27 April 2007
+ m68k version finished

28 April 2007
+ Optimized m68k version

RELEASE 0.25

6 July 2007
+ Have VAX "hello world" stub going
+ Also have avr32 "hello world" going

8 July 2007
+ avr32 working
+ super-optimized avr32 and got it to 914 bytes, a new record!


ll.ppc.s
#
#  linux_logo in ppc assembler    0.38
#
#  by Vince Weaver <vince _at_ deater.net>
#
#  assemble with     "as -o ll.o ll.ppc.s"
#  link with         "ld -o ll ll.o"

#  BUGS:  No Pretty Printing (rounding, cpu-model cleanup)
#      :  Doesn't print vendor name
#      :  doesn't count CPUs on SMP systems


#  notes ...
#    a lot of the offsets are off by 1.  This is to use the more
#      efficient lbu commands that auto-update.
#    don't use non callee saved registers around a syscall.
#      it looks like Linux does not save those anymore

# Register        Usage                CALLEE SAVE
# r0            prolog/epilog           NO
# r1            stack pointer           YES
# r2            TOC pointer (reserved)  YES
# r3-r4         1/2 para and return     NO
# r5-r10        3-8th para              NO
# r11-r12       Func Linkage reg        NO
# r12           Used by global linkage  NO
# r13           Small data area pointer NO
# r14-r30       General Int registers   YES
# r31           Global Environment Ptr  YES	  
# f0            Scratch                 NO
# f1            1st param/return        NO
# f2-f8         2-8th fp param          NO
# f9-f13        Scratch Reg             NO
# f14-f31       Global fp regs          YES
# CR0-CR7       Condition Regs          2,3,4 Yes
# LR            Link register           YES
# CTR           Counter register        NO
# XER           Fixed Point Exception   NO
# FPSCR         fp status& ctrl         NO

# CR = condition register.
#   There are actually 8 condition registers, CR0-CR7
# CR0, default for "."  =  4 bits.  Negative, Positive, Zero, Overflow
# CR1, gets floating point results
# CRx, as result of compare instruction.  
#      bit 0=less than, 1=greater than 2=equal, 3=overflow
# XER = holds carry and overflow flags
# CTR = counter register, used as a loop index

# Instruction set
# add add. addo addo. - "." means update condition reg. "o" means handle oflo
# addc addc. addco addco. - add while setting carry
# addex adde. addeo addeo. - add extended (with carry from XER)
# addi - add immediate
# addic addic. - add immediate while setting carry
# addis - add immediate shifted (high bits from immediate << 16)
# addme addme. addmeo addmeo. - add to minus one extended  (ra + carry + -1)
# addze addze. addzeo addzeo. - add to zero extended  (ra + carry)
# and and. - logical and
# andc andc. - and compliment
# andi. - and immediate
# andis. - and immediate shifted << 16
# b ba bl bla - a means absolute, l means and link
# bc bca bcl bcla - branch conditional.  bc BO,BI,target
#                   BO specifies true, false, always, CTR
#                   BI specifies bit in the CR to use
# bcctr bcctrl - branch conditional to count register
# bclr bclrl - branch conditional to link register
# cmp crfD,L,rA,rB - compare.  crfD = which CR to use, L is 32 or 64 bit?
# cmpi - compare immediate
# cmpl - compare logical (unsigned?)
# cmpli - compare logical immediate
# cntlzw cntlzw. - count leading zeros
# creqv - condition register equivelent
# crand crnand crandc - condition register and nand and compliment
# cror crnor crorc - condition register or nor or compliment
# crxor - condition register xor
# divw divw. divwo divwo. - divide word
#    no remainder.. need to divw/mullw/subf to get it
# divwu divwu. divwuo divwuo. - divide unsigned
# eieio - enforce in-order execution of external io
# eqv eqv. - exclusive nor
# extsb extsb. - extend sign byte
# extsh extsh. - estend sign half
# lbz - load byte and zero
# lbzu - load byte and update.  the Effective address is stored back into RA
# lbzux - load byte and update, indexed.  EA is RA+RB
# lbzx - load byte indexed
# lha lhau - load half and sign extend, with update
# lhax, laux - load half indexed, sign extend, with update
# lhbrx - load half byte reversed indexed
# lhz lhzu lhzux lhzx - load half all the various ways
# lmw - load multiple words.  multiple words loaded into consecutive regs
# lswi lswx - load string word immediate.  load bytes into regs
# lwarx - load word and reserve (for atomic)
# lwbrx - load word byte reversed
# lwz lwzu lwzux lwzx - load word and zero
# mcrf - move cond reg field
# mcrxr - move xer to cond reg
# mfcr - move from cond reg
# mulhw mulhw. - multiply high word - 32x32, get resulting top 32 bits
# mulhwu mulhwu. - multiplu high word unsigned
# mulli - multiply low immediate
# mullw mullw. mullwo mullwo. - muliply low
# nand nand. - nand
# neg neg. nego nego. - negate
# nor nor. - nor
# or or. - or
# orc orc. - or with complement
# ori oris - or immediate, or immediate shifted
# rlwimi rlwmimi. - ra,rs,sh,mb,me - rotate left word immed then mask insert
#                   rs rotated by sh. mb to me specify mask
# rlwinm rlwinm. - rotate left word immed then and mask
# rlwnm rlwnm. - rotate left word and with mask
# sc - system call
# slw slw. - shift left word 
# sraw sraw. - shift right algebraic word
# srawi srawi. - shift right algebraic word immediate
# srw srw. - shift right word
# stb stbu stbux stbx - store byte, update, indexed
# sth sthu sthux sthx - store halfword
# sthbrx - store halfword byte reversed
# stmw - store multiple regs
# stswi stswx - store string word immediate, indexed
# stw stwu stwux stwx - store word, update
# stwbrx - store word byte reversed indexed
# stwcx. - store word conditional (for atomic)
# subf subf. subfo subfo. - subtract from
# subfc subfc. subfco subfco. - subtract from carrying
# subfe subfe. subfeo subfeo. - subtract from extended (carry)
# subfic - sub from immediate carrying
# subfme subfme. subfmeo subfmeo. - subtract from minus one extended
# subfze subfze. subfzeo subfzeo. - subtract from zero extended
# xor xor. - xor
# xori xoris - xor immediate, shifted

# Pseudo instructions
#  li, la, subi
#  blt, bne cr2, bdnz
#  bltctr, bnectr cr2
#  bltlr, bnelr cr2, bdnzlr
#  cmpd, cmpw cr3
#  cmpid, cmpiw
#  cmpld, cmplw
#  cmpldi, cmplwi
#  crse, crclr
#  not
#  mr - move register
#  nop
#  extlwi extrwi rotlwi rotrwi slwi srwi clrlwi clrrwi clrlslwi
#  rotlw
#  sub subc

# Optimization:
# + 1237 - already optimized version on old power3 machine
# + 1225 - use lhbrx, remove unneeded var copying
# + 1221 - use cmpwi in find_string
# + 1217 - use subfic
# + 1205 - have center_and_print load strings based on DATA_BEGIN
# + 1201 - change arguments to num_to_ascii to avoid copy
# + 1197 - only restore link register once
# + 1165 - compile with a modern toolchain (2.19)

# offsets into the results returned by the uname syscall
.equ U_SYSNAME,0
.equ U_NODENAME,65
.equ U_RELEASE,65*2
.equ U_VERSION,(65*3)
.equ U_MACHINE,(65*4)
.equ U_DOMAINNAME,65*5

# offset into the SYSCALL_SYSINFO buffer
.equ S_TOTALRAM,16

# Sycscalls
.equ SYSCALL_EXIT,     1
.equ SYSCALL_READ,     3
.equ SYSCALL_WRITE,    4
.equ SYSCALL_OPEN,     5
.equ SYSCALL_CLOSE,    6
.equ SYSCALL_SYSINFO,116
.equ SYSCALL_UNAME,  122

#
.equ STDIN, 0
.equ STDOUT,1
.equ STDERR,2

.equ BSS_BEGIN,25
.equ DATA_BEGIN,26

.include "logo.include"

	.globl _start	
_start:	

        #========================
	# Initialization
	#========================
	

#	eieio				# coolest opcode of all time ;)
					# not needed, but I had to put it here
  	# the hack loading BSS_BEGIN and DATA_BEGIN
	# saves one instruction on any future load from memory
	# as we can just do an addi rather than an lis;addi

	lis	25,bss_begin@ha
	addi	25,25,bss_begin@l	# bss offset in r25
	
	lis	26,data_begin@ha
	addi	26,26,data_begin@l	# data offset in r26

	addi	17,BSS_BEGIN,(out_buffer-bss_begin)
					# output buffer in r17

	addi	21,BSS_BEGIN,(text_buf-bss_begin)
					# text_buf in r21
 	     	

        #=========================
	# PRINT LOGO
	#=========================

# LZSS decompression algorithm implementation
# by Stephan Walter 2002, based on LZSS.C by Haruhiko Okumura 1989
# optimized some more by Vince Weaver


	li	8,(N-F)			# grab "R"

	addi	9,DATA_BEGIN,(logo-data_begin)-1
					# logo_pointer

	addi	12,DATA_BEGIN,(logo_end-data_begin)-1
					# end of the logo
					
	mr      16,17			# r16 starts at output pointer

decompression_loop:
	lbzu 	10,1(9)			# load in a byte
					# auto-update
	ori	11,10,0xff00		# load top as a hackish 
					# 8-bit counter

test_flags:
	cmpw	0,12,9			# have we reached the end?
	ble	done_logo		# if so exit

	andi.	13,11,0x1		# test bottom bit
	srawi   11,11,1			# shift over
	
	bne	0,discrete_char		# if bit was 0, we have a single char

offset_length:
	      				# Load 16-bit little endian
	addi	9,9,1			# have to do this because
					# we use lbzu for rest of loads
	lhbrx	24,0,9			# load half-word byte reversed index
					#   0 means use 0, not r0
	addi	9,9,1			# we loaded two bytes

	srawi  15,24,P_BITS
	addi   15,15,THRESHOLD+1 	# match length is top bits
	       			 
output_loop:
	andi.  24,24,(POSITION_MASK<<8+0xff)	
	       				# offset in text_buf is bottom bits
					
	lbzx   10,21,24			# load byte from text_buf	
	addi   24,24,1			# increment pointer
	
store_byte:
	stbu   10,1(16)			# store byte to putput
	
	stbx    10,21,8			# store byte to text_buf
	addi	8,8,1			# increment pointer
	andi.	8,8,(N-1)		# mask to prevent overflow

	addic.	15,15,-1		# decrement count
	bne	0,output_loop		# loop if not output
	
	andi.	13,11,0xff00		# test to see if done with 8 bits of
	bne	test_flags		# flags
	
	b	decompression_loop	# loop

discrete_char:

	lbzu    10,1(9)			# load byte to output
	li	15,1			# set to only output one byte

	b       store_byte		# go to store byte routine

done_logo:

	addi	4,17,1			# restore output pointer
					# (plus one because r17 is decremented)
	bl	write_stdout		# and print the logo
	

        #==========================
	# First Line
	#==========================

first_line:
	mr	14,17		    	# copy output pointer to r14

	#==========================
	# PRINT VERSION
	#==========================

run_uname:

	li	0,SYSCALL_UNAME		# uname syscall
	addi	3,BSS_BEGIN,(uname_info-bss_begin)		
					# uname struct
	sc				# do syscall

print_version:

	addi	16,BSS_BEGIN,(uname_info-bss_begin)+U_SYSNAME@l-1	
					# os-name from uname "Linux"
	bl	strcat
	
	addi	16,DATA_BEGIN,(ver_string-data_begin)-1
					# source is " Version "
	bl 	strcat
	
	addi	16,BSS_BEGIN,(uname_info-bss_begin)+U_RELEASE@l-1
					# version from uname "2.4.1"
	bl 	strcat
	
	addi	16,DATA_BEGIN,(compiled_string-data_begin)-1
					# source is ", Compiled "
	bl 	strcat

	addi	16,BSS_BEGIN,(uname_info-bss_begin)+U_VERSION-1
      					# compiled date
	bl 	strcat
	
	bl	center_and_print	# write it to screen
	

	#===============================
	# Middle-Line
	#===============================
middle_line:

	mr	14,17 			# point output to out_buf

	#=========
	# Load /proc/cpuinfo into buffer
	#=========

	li	0,SYSCALL_OPEN		# open()
	addi	3,DATA_BEGIN,(cpuinfo-data_begin)		
					# '/proc/cpuinfo'
	li	4,0			# O_RDONLY <bits/fcntl.h>
	sc				# syscall.  fd in r0.  
					# we should check that r0>=0
					
	mr	13,3			# save fd in r13
	
	li	0,SYSCALL_READ		# read
	addi	4,BSS_BEGIN,(disk_buffer-bss_begin)
	li	5,4096		 	# assume cpuinfo file < 4k
	sc	

	mr	3,13			# restore fd
	li	0,6			# close
	sc

	#=============
	# Number of CPUs
	#=============
	
num_cpu:
	# Assume 1 CPU for now
	# my iBook's /proc/cpuinfo does not have a "processor" line ???
	
	addi	16,DATA_BEGIN,(one-data_begin)-1
	bl	strcat
	
	#=========
	# MHz
	#=========

mhz:
    	lis	20,('l'<<8)+'o'		# find 'lock ' and grab up to M
	addi	20,20,('c'<<8)+'k'
	li	23,'M'			
   	bl	find_string
   
	addi	16,DATA_BEGIN,(megahertz-data_begin)-1
					# print 'MHz '
	bl	strcat
   
  
	#=========
	# Chip Name
	#=========
chip_name:	
   	lis     20,('c'<<8)+'p'     	# find 'cpu\t: ' and grab up to \n
	addi	20,20,('u'<<8)+'\t'
	li	23,'\n'
	bl	find_string
	
	addi	16,DATA_BEGIN,(comma-data_begin)-1
					# print ', '
	bl	strcat
	
	#========
	# RAM
	#========
ram:	
	li	0,SYSCALL_SYSINFO	# sysinfo() syscall
	addi	3,BSS_BEGIN,(sysinfo_buff-bss_begin)
					# sysinfo_buffer

	sc

	lwz	4,(sysinfo_buff+S_TOTALRAM-bss_begin)(BSS_BEGIN)
					# load bytes of RAM into r4

	srwi	19,4,20		# divide by 2^20 to get MB
	
	li	5,0

	bl	num_to_ascii

	addi	16,DATA_BEGIN,(ram_comma-data_begin)-1
					# print 'M RAM, '

	bl	strcat
	
	#========
	# Bogomips
	#========

bogomips:
	lis	20,('m'<<8)+'i'		# find 'mips' and grab up to \n
	addi	20,20,('p'<<8)+'s'
	li	23,'\n'
	bl	find_string
      
	addi	16,DATA_BEGIN,(bogo_total-data_begin)-1
					# print "Bogomips Total"
	bl	strcat

	bl	center_and_print	# center it


	#=================================
	# Print Host Name
	#=================================
host_name:	
	mr	14,17			# restore out buffer
	
	addi	16,BSS_BEGIN,((uname_info-bss_begin)+U_NODENAME)-1
					# hostname		      
					
	bl	strcat				
	
	bl	center_and_print

	addi	4,DATA_BEGIN,(default_colors-data_begin)
					# restore default colors
					
	bl	write_stdout				

	#================================
	# Exit
	#================================
exit:	
        li      3,0		# 0 exit value
	li      0,SYSCALL_EXIT  # put the exit syscall number in eax
	sc	             	# and exit


	#=================================
	# FIND_STRING 
	#=================================
	#   r23 is char to end at
	#   r20 is the 4-char ascii string to look for
	#   r14 points at output buffer
	#   r16,r21

find_string:
		
	addi	16,BSS_BEGIN,(disk_buffer-bss_begin)-1	
					# look in cpuinfo buffer
					# -1 so we can use lbzu
	
find_loop:
	lwzu	13,1(16)		# load in 32 bits, incrementing 8bits
	cmpwi	13,0			# if null, we are done
	beq	done
	cmpw	13,20			# compare with out 4 char string
	bne	find_loop		# if no match, keep looping

	
					# if we get this far, we matched
					
find_colon:
	lbzu	13,1(16)		# repeat till we find colon
	
	cmpwi	13,0			# if zero, bail
	beq	done
	
	cmpwi	13,':'			# compare to colon
	bne	find_colon

	addi	16,16,1			# skip a char [should be space]
	
store_loop:	 
	 lbzu	13,1(16)		# load byte
	 
	 cmpwi	13,0			# if zero, bail
	 beq	done
	 
    	 cmpw	13,23			# is it end string?
	 beq 	almost_done		# if so, finish
	 stbu	13,1(14)		# if not store and continue
	 b	store_loop
	 
almost_done:	 
	li	13,0			# replace last value with null
	stb	13,1(14)

done:
	blr				# return

	#==============================
	# center_and_print
	#==============================
	# r14 is end of buffer
	# r17 is start of buffer
	# r29 = saved link register
	# r4-r10, r19-r22, r30 trashed
	
center_and_print:

	mflr 	29			# back up return address

	subf	5,17,14			# see how long the output
					# buffer is
					
	cmpwi	5,80			# see if we are >80
        bgt	done_center		# if so, bail

	subfic	4,5,80			# r4 = 80-r5
					#   is it possible to combine this
					#   with the >80 test?
					
	srawi	23,4,1			# divide by two

	addi	4,DATA_BEGIN,(escape-data_begin)
	bl	write_stdout

	mr	19,23	    		# move size into argument
	li	5,1			# print to stdout
	bl	num_to_ascii		# print number
	
	addi	4,DATA_BEGIN,(c-data_begin)
	bl	write_stdout

done_center:	

	addi	4,17,1			# move string to output+1
	bl	write_stdout		# call write stdout

	addi	4,DATA_BEGIN,(linefeed-data_begin)

	mtlr	29	      		# restore link register
					# and let write_stdout
					# return for us



	#================================
	# WRITE_STDOUT
	#================================
	# r4 has string
	# r0,r3,r4,r5,r6 trashed
		
write_stdout:
	li	0,SYSCALL_WRITE		# write syscall
	li	3,STDOUT		# stdout	
	
	li	5,0			# string length counter
strlen_loop:
	lbzx 	6,4,5			# get byte from (r4+r5)
       	addi	5,5,1			# increment counter
	cmpi	0,6,0			# is it zero?
	bne	strlen_loop		# if not keep counting
	
	addi	5,5,-1			# adjust back down
	sc				# syscall
	
	blr				# return


	#================================
	# Num to Ascii
	#================================
	# num is in r19
	#  breaks on negative values
	# r5 =0 then strcat, otherwise stdout
	# r5-r10,r19,r20,r21,r22,r30 trashed	

num_to_ascii:

	mflr    30			# save the link register

	addi	16,BSS_BEGIN,(num_to_ascii_end-bss_begin)
					# the end of a backwards growing
					# 10 byte long buffer.  
					
	li	20,10			# we will divide by 10
	
div_by_10:
	divw	21,19,20		# divide r19 by r20 put into r21 
	
	mullw	22,21,20		# find remainder.  1st q*dividend
	subf	22,22,19		# then subtract from original = R
	addi	22,22,0x30		# convert remainder to ascii
    	
	stbu	22,-1(16)		# Store to backwards buffer
	
	mr	19,21			# move Quotient as new dividend
	cmpwi	19,0			# was quotient zero?
	bne    	div_by_10		# if not keep dividing
	
write_out:
	mtlr	30			# restore link register

	cmpwi	5,0			# if r5 is 0 then skip ahead
	beq 	strcat_num		

stdout_num:
        mr	4,16			# point to our buffer
	b	write_stdout		# stdout will return for us

strcat_num:
	addi	16,16,-1		# point to the beginning

					# fall through to strcat
					
	#================================
	# strcat
	#================================
	# r13 = "temp"
	# r16 = "source"
       	# r14 = "destination"
strcat:
	lbzu	13,1(16)		# load a byte from [r16]
	stbu	13,1(14)		# store a byte to [r14]
	cmpwi	13,0			# is it zero?
	bne	strcat			# if not loop
	subi	14,14,1			# point to one less than null
	blr				# return


#===========================================================================
.data
#===========================================================================


data_begin:

.include "logo.lzss_new"

ver_string:	.ascii	" Version \0"
compiled_string:	.ascii	", Compiled \0"
megahertz:	.ascii	"MHz PPC \0"
.equ comma, ram_comma+5
linefeed:   	.ascii  "\n\0"
escape:		.ascii	"\033[\0"
c:		.ascii  "C\0"
ram_comma:	.ascii	"M RAM, \0"

bogo_total:	.ascii	" Bogomips Total\0"

default_colors:	.ascii	"\033[0m\n\0"

.ifdef FAKE_PROC
cpuinfo:	.ascii	"proc/cpui.ppc\0"
.else
cpuinfo:	.ascii	"/proc/cpuinfo\0"
.endif

one:	.ascii	"One \0"

#============================================================================
#.bss
#============================================================================

.lcomm bss_begin,0
.lcomm	num_to_ascii_buff,10
.lcomm num_to_ascii_end,1
.lcomm  text_buf, (N+F-1)	# These buffers must follow each other
.lcomm	out_buffer,16384

	# see /usr/src/linux/include/linux/kernel.h
	
.lcomm  sysinfo_buff,(64)
.lcomm  uname_info,(65*6)

.lcomm	disk_buffer,4096,4	# we cheat!!!!




